### PR TITLE
Blackparrot with opensbi v1.0

### PR DESCRIPTION
**Pull request by Yuan-Mao**
This PR brings initial support for Opensbi version 1.0.

I have tested it on Dromajo with Linux kernel 5.15 with OPENSBI_NCPUS set to 4. Nearly all the changes are based on what we have done in the branch 'blackparrot_mods'.

```

OpenSBI v1.0                                                                                                                                                                                               
   ____                    _____ ____ _____                                                                                                                                                                
  / __ \                  / ____|  _ \_   _|                                                                                                                                                               
 | |  | |_ __   ___ _ __ | (___ | |_) || |                                                                                                                                                                 
 | |  | | '_ \ / _ \ '_ \ \___ \|  _ < | |                                                                                                                                                                 
 | |__| | |_) |  __/ | | |____) | |_) || |_                                                                                                                                                                
  \____/| .__/ \___|_| |_|_____/|____/_____|                                                                                                                                                               
        | |                                                                                                                                                                                                
        |_|                                                                                                                                                                                                
                                                                                                                                                                                                           
Platform Name             : BlackParrot                                                                                                                                                                    
Platform Features         : medeleg                                                                                                                                                                        
Platform HART Count       : 4                                                                                                                                                                              
Platform IPI Device       : aclint-mswi                                                                                                                                                                    
Platform Timer Device     : aclint-mtimer @ 10000000Hz                                                                                                                                                     
Platform Console Device   : bp-console                                                                                                                                                                     
Platform HSM Device       : ---                                                                                                                                                                            
Platform Reboot Device    : ---                                                                                                                                                                            
Platform Shutdown Device  : bp-reset                                                                                                                                                                       
Firmware Base             : 0x80000000                                                                                                                                                                     
Firmware Size             : 240 KB                                                                                                                                                                         
Runtime SBI Version       : 0.3                                                                                                                                                                            
                                                                                                                                                                                                           
Domain0 Name              : root                                                                                                                                                                           
Domain0 Boot HART         : 3                                                                                                                                                                              
Domain0 HARTs             : 0*,1*,2*,3*                                                                                                                                                                    
Domain0 Region00          : 0x0000000000308000-0x000000000030bfff (I)                                                                                                                                      
Domain0 Region01          : 0x0000000000300000-0x0000000000307fff (I)                                                                                                                                      
Domain0 Region02          : 0x0000000080000000-0x000000008003ffff ()                                                                                                                                       
Domain0 Region03          : 0x0000000000000000-0xffffffffffffffff (R,W,X)                                                                                                                                  
Domain0 Next Address      : 0x0000000080200000                                                                                                                                                             
Domain0 Next Arg1         : 0x0000000080040000                                                                                                                                                             
Domain0 Next Mode         : S-mode                                                                                                                                                                         
Domain0 SysReset          : yes                                                                                                                                                                            
                                                                                                                                                                                                           
Boot HART ID              : 3                                                                                                                                                                              
Boot HART Domain          : root                                                                                                                                                                           
Boot HART ISA             : rv64imafdsu                                                                                                                                                                    
Boot HART Features        : scounteren,mcounteren,mcountinhibit                                                                                                                                            
Boot HART PMP Count       : 0                                                                                                                                                                              
Boot HART PMP Granularity : 0                                                                                                                                                                              
Boot HART PMP Address Bits: 0                                                                                                                                                                              
Boot HART MHPM Count      : 0                                                                                                                                                                              
Boot HART MIDELEG         : 0x0000000000000222                                                                                                                                                             
Boot HART MEDELEG         : 0x000000000000b109          
```

### CODE REVIEW

**Comment by dpetrisko** [comment]
Awesome! @farzamgl can you check this out?

**Comment by farzamgl** [comment]
Looks good! @Yuan-Mao any new or dropped patches compared to the blackparrot_mods branch?

**Comment by Yuan-Mao** [comment]
Hi @farzamgl 
No, I pretty much followed the modification from 'blackparrot_mods'. Nothing significant added or removed.

The only thing that caught me for a little while was the introducing of mtimer and mswi devices in opensbi v1.0, but those were just basically some code refactoring of some previous code in template/platform.c, so I just copied and used the code from blackparrot/platform.c and they should be fine.

File: firmware/fw_base.S [hunk 1]
```diff
@@ -58,10 +59,10 @@ _start:
 	bne	a0, a6, _wait_relocate_copy_done
 _try_lottery:
 	/* Jump to relocation wait loop if we don't get relocation lottery */
-	lla	a6, _relocate_lottery
-	li	a7, 1
-	amoadd.w a6, a7, (a6)
-	bnez	a6, _wait_relocate_copy_done
+	lla	a0, _relocate_lottery
+	li	a1, 1
+	call    amo_addw
```
Comment (dpetrisko) [inline firmware/fw_base.S]: Should be able to revert these changes now that we support L1 AMOs natively

**Comment by dpetrisko** [comment]
This is kind of hard to review against master since it's all BP changes, but diffing against blackparrot_mods gives us 1000s of commits where master has been updated. Probably the best thing is @farzamgl to review it and then test it on a board with linux 5.15 and then we can merge https://github.com/black-parrot-sdk/bp-linux/pull/6 as well

**Comment by farzamgl** [comment]
The mods look good. I'll try it on the board tomorrow.

**Comment by farzamgl** [comment]
@dpetrisko @Yuan-Mao Boots on Ultra96v2.

**Comment by dpetrisko** [comment]
Awesome, thanks guys!
